## Segmenting and Clustering Neighborhoods in Toronto

Author: Shahid Mian
Dated: January 22, 2021
Notebook contains Questions 1, 2 & 3 of the Assignment


In [1]:
import pandas as pd


Question # 1

Impoprting Data


In [2]:
import requests

In [3]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
Wiki_url = requests.get(url)
Wiki_url

<Response [200]>

In [4]:
Wiki_data = pd.read_html(Wiki_url.text)
Wiki_data

[    Postal Code           Borough  \
 0           M1A      Not assigned   
 1           M2A      Not assigned   
 2           M3A        North York   
 3           M4A        North York   
 4           M5A  Downtown Toronto   
 ..          ...               ...   
 175         M5Z      Not assigned   
 176         M6Z      Not assigned   
 177         M7Z      Not assigned   
 178         M8Z         Etobicoke   
 179         M9Z      Not assigned   
 
                                          Neighbourhood  
 0                                         Not assigned  
 1                                         Not assigned  
 2                                            Parkwoods  
 3                                     Victoria Village  
 4                            Regent Park, Harbourfront  
 ..                                                 ...  
 175                                       Not assigned  
 176                                       Not assigned  
 177                

In [5]:
len(Wiki_data), type(Wiki_data)

(3, list)

In [6]:
Wiki_data = Wiki_data[0]
Wiki_data

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                         Neighbourhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

In [7]:
df = Wiki_data[Wiki_data["Borough"] != "Not assigned"]
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [8]:
df = df.groupby(['Postal Code']).head()
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [9]:
df.Neighbourhood.str.count("Not assigned").sum()

0

In [10]:
df = df.reset_index()
df

index Postal Code           Borough  \
0        2         M3A        North York   
1        3         M4A        North York   
2        4         M5A  Downtown Toronto   
3        5         M6A        North York   
4        6         M7A  Downtown Toronto   
..     ...         ...               ...   
98     160         M8X         Etobicoke   
99     165         M4Y  Downtown Toronto   
100    168         M7Y      East Toronto   
101    169         M8Y         Etobicoke   
102    178         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 4 columns]

In [11]:
df.drop(['index'], axis = 'columns', inplace = True)
df

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [12]:
df.shape

(103, 3)

Anwser Question # 1 : We have 103 Rows and 3 Columns eg. Postal Code, Borough and Neighbourhood ..


## Submit Above link to your Notebook on your Github repository (10 marks)


Question # 2

In [13]:
# pip install geocoder

In [14]:
import geocoder # import geocoder

In [15]:
data = pd.read_csv("https://cocl.us/Geospatial_data")
data

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

In [16]:
print("The Shape of our Wiki data is: ", df.shape)
print("the Shape of our csv data is: ", data.shape)

The Shape of our Wiki data is:  (103, 3)
the Shape of our csv data is:  (103, 3)


In [17]:
df.dtypes

Postal Code      object
Borough          object
Neighbourhood    object
dtype: object

In [18]:
data.dtypes

Postal Code     object
Latitude       float64
Longitude      float64
dtype: object

In [19]:
combined_data = df.join(data.set_index('Postal Code'), on='Postal Code', how='inner')
combined_data

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100  Business reply mail Processing Centre, South C...  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

In [20]:
combined_data.shape

(103, 5)

Solution: We get 103 rows as expected when we do a inner join, so we have good data.

Question 3:
Drawing inspiration from the previous lab where we cluster the neighbourhood of NYC, We cluster Toronto based on the similarities of the venues categories using Kmeans clustering and Foursquare API.

In [21]:
from geopy.geocoders import Nominatim

In [22]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The coordinates of Toronto are 43.6534817, -79.3839347.


In [23]:
#!pip3 install folium
import folium

In [24]:
# Creating Toronto Map
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# adding markers
for latitude, longitude, borough, neighbourhood in zip(combined_data['Latitude'], combined_data['Longitude'], combined_data['Borough'], combined_data['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='red',
        fill=True
        ).add_to(map_Toronto)  
    
map_Toronto

Foursquare API credentials

In [25]:
CLIENT_ID = 'KLOCDNEAHS3Q1V0JGTXF4FXF0PLTBFNNJ3WI5ACUGCYJD0YR'
CLIENT_SECRET = 'YCAAAQIGGFGUUS24DQ5YHMWIGYXKKJXHVVS14G2QV1I5IKZB'
VERSION = '20180605' # Foursquare API version
print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: KLOCDNEAHS3Q1V0JGTXF4FXF0PLTBFNNJ3WI5ACUGCYJD0YR
CLIENT_SECRET:YCAAAQIGGFGUUS24DQ5YHMWIGYXKKJXHVVS14G2QV1I5IKZB


In [26]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)       
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius
            )
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['categories'][0]['name']) for v in results])
    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Category']
    return(nearby_venues)

In [27]:
venues_in_toronto = getNearbyVenues(combined_data['Neighbourhood'], combined_data['Latitude'], combined_data['Longitude'])

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

In [28]:
venues_in_toronto.shape

(1329, 5)

In [29]:
venues_in_toronto.head()

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0         Parkwoods               43.753259               -79.329656   
1         Parkwoods               43.753259               -79.329656   
2         Parkwoods               43.753259               -79.329656   
3  Victoria Village               43.725882               -79.315572   
4  Victoria Village               43.725882               -79.315572   

                    Venue         Venue Category  
0         Brookbanks Park                   Park  
1     Towns On The Ravine                  Hotel  
2           Variety Store      Food & Drink Shop  
3  Victoria Village Arena           Hockey Arena  
4               Portugril  Portuguese Restaurant

In [30]:
venues_in_toronto.groupby('Neighbourhood').head()

Neighbourhood  \
0                                             Parkwoods   
1                                             Parkwoods   
2                                             Parkwoods   
3                                      Victoria Village   
4                                      Victoria Village   
...                                                 ...   
1313  Mimico NW, The Queensway West, South of Bloor,...   
1314  Mimico NW, The Queensway West, South of Bloor,...   
1315  Mimico NW, The Queensway West, South of Bloor,...   
1316  Mimico NW, The Queensway West, South of Bloor,...   
1317  Mimico NW, The Queensway West, South of Bloor,...   

      Neighbourhood Latitude  Neighbourhood Longitude                   Venue  \
0                  43.753259               -79.329656         Brookbanks Park   
1                  43.753259               -79.329656     Towns On The Ravine   
2                  43.753259               -79.329656           Variety Store   
3                  43.725882               -79.315572  Victoria Village Arena   
4                  43.725882               -79.315572               Portugril   
...                      ...                      ...                     ...   
1313               43.628841               -79.520999              Wingporium   
1314               43.628841               -79.520999        South St. Burger   
1315               43.628841               -79.520999               Dollarama   
1316               43.628841               -79.520999          Healthy Planet   
1317               43.628841               -79.520999    Artisano Bakery Café   

             Venue Category  
0                      Park  
1                     Hotel  
2         Food & Drink Shop  
3              Hockey Arena  
4     Portuguese Restaurant  
...                     ...  
1313            Wings Joint  
1314           Burger Joint  
1315         Discount Store  
1316        Supplement Shop  
1317                 Bakery  

[418 rows x 5 columns]

Note: There are 418 Records for each Neighbourhood

Now checking Neighbourhood Categories

In [31]:
venues_in_toronto.groupby('Venue Category').max()

Neighbourhood  \
Venue Category                                                          
Accessories Store                                   Wexford, Maryvale   
Adult Boutique                                   Church and Wellesley   
Airport                                                     Downsview   
Airport Food Court  CN Tower, King and Spadina, Railway Lands, Har...   
Airport Gate        CN Tower, King and Spadina, Railway Lands, Har...   
...                                                               ...   
Wine Bar                     Toronto Dominion Centre, Design Exchange   
Wine Shop                                Dufferin, Dovercourt Village   
Wings Joint         Mimico NW, The Queensway West, South of Bloor,...   
Women's Store                                     Caledonia-Fairbanks   
Yoga Studio                            University of Toronto, Harbord   

                    Neighbourhood Latitude  Neighbourhood Longitude  \
Venue Category                                                        
Accessories Store                43.750072               -79.295849   
Adult Boutique                   43.665860               -79.383160   
Airport                          43.737473               -79.394420   
Airport Food Court               43.628947               -79.394420   
Airport Gate                     43.628947               -79.394420   
...                                    ...                      ...   
Wine Bar                         43.653206               -79.379817   
Wine Shop                        43.669005               -79.442259   
Wings Joint                      43.628841               -79.520999   
Women's Store                    43.689026               -79.453512   
Yoga Studio                      43.715383               -79.340923   

                                              Venue  
Venue Category                                       
Accessories Store                       Puffin Gear  
Adult Boutique                            Seduction  
Airport             Toronto Downsview Airport (YZD)  
Airport Food Court                Billy Bishop Café  
Airport Gate                                 Gate 8  
...                                             ...  
Wine Bar                          The National Club  
Wine Shop                                    Macedo  
Wings Joint                              Wingporium  
Women's Store                         Maximum Woman  
Yoga Studio                               YogaSpace  

[236 rows x 4 columns]

Note: There are around 236 different types of Venue Categories in Toronto

One Hot encoding the venue Categories

In [32]:
toronto_venue_cat = pd.get_dummies(venues_in_toronto[['Venue Category']], prefix="", prefix_sep="")
toronto_venue_cat

Accessories Store  Adult Boutique  Airport  Airport Food Court  \
0                     0               0        0                   0   
1                     0               0        0                   0   
2                     0               0        0                   0   
3                     0               0        0                   0   
4                     0               0        0                   0   
...                 ...             ...      ...                 ...   
1324                  0               0        0                   0   
1325                  0               0        0                   0   
1326                  0               0        0                   0   
1327                  0               0        0                   0   
1328                  0               0        0                   0   

      Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0                0               0                0                 0   
1                0               0                0                 0   
2                0               0                0                 0   
3                0               0                0                 0   
4                0               0                0                 0   
...            ...             ...              ...               ...   
1324             0               0                0                 0   
1325             0               0                0                 0   
1326             0               0                0                 0   
1327             0               0                0                 0   
1328             0               0                0                 0   

      American Restaurant  Antique Shop  ...  Vegetarian / Vegan Restaurant  \
0                       0             0  ...                              0   
1                       0             0  ...                              0   
2                       0             0  ...                              0   
3                       0             0  ...                              0   
4                       0             0  ...                              0   
...                   ...           ...  ...                            ...   
1324                    0             0  ...                              0   
1325                    0             0  ...                              0   
1326                    0             0  ...                              0   
1327                    0             0  ...                              0   
1328                    0             0  ...                              0   

      Video Game Store  Video Store  Vietnamese Restaurant  Warehouse Store  \
0                    0            0                      0                0   
1                    0            0                      0                0   
2                    0            0                      0                0   
3                    0            0                      0                0   
4                    0            0                      0                0   
...                ...          ...                    ...              ...   
1324                 0            0                      0                0   
1325                 0            0                      0                0   
1326                 0            0                      0                0   
1327                 0            0                      0                0   
1328                 0            0                      0                0   

      Wine Bar  Wine Shop  Wings Joint  Women's Store  Yoga Studio  
0            0          0            0              0            0  
1            0          0            0              0            0  
2            0          0            0              0            0  
3            0          0            0              0            0  
4            0       

In [33]:
toronto_venue_cat['Neighbourhood'] = venues_in_toronto['Neighbourhood'] 

# moving neighborhood column to the first column
fixed_columns = [toronto_venue_cat.columns[-1]] + list(toronto_venue_cat.columns[:-1])
toronto_venue_cat = toronto_venue_cat[fixed_columns]

toronto_venue_cat.head()

Neighbourhood  Accessories Store  Adult Boutique  Airport  \
0         Parkwoods                  0               0        0   
1         Parkwoods                  0               0        0   
2         Parkwoods                  0               0        0   
3  Victoria Village                  0               0        0   
4  Victoria Village                  0               0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  ...  Vegetarian / Vegan Restaurant  \
0                 0                    0  ...                              0   
1                 0                    0  ...                              0   
2                 0                    0  ...                              0   
3                 0                    0  ...                              0   
4                 0                    0  ...                              0   

   Video Game Store  Video Store  Vietnamese Restaurant  Warehouse Store  \
0                 0            0                      0                0   
1                 0            0                      0                0   
2                 0            0                      0                0   
3                 0            0                      0                0   
4                 0            0                      0                0   

   Wine Bar  Wine Shop  Wings Joint  Women's Store  Yoga Studio  
0         0          0            0              0            0  
1         0          0            0              0            0  
2         0          0            0              0            0  
3         0          0            0              0            0  
4         0          0            0              0            0  

[5 rows x 237 columns]

In [34]:
toronto_grouped = toronto_venue_cat.groupby('Neighbourhood').mean().reset_index()
toronto_grouped.head()


Neighbourhood  Accessories Store  \
0                                        Agincourt                0.0   
1                           Alderwood, Long Branch                0.0   
2  Bathurst Manor, Wilson Heights, Downsview North                0.0   
3                                  Bayview Village                0.0   
4                Bedford Park, Lawrence Manor East                0.0   

   Adult Boutique  Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0             0.0      0.0                 0.0           0.0             0.0   
1             0.0      0.0                 0.0           0.0             0.0   
2             0.0      0.0                 0.0           0.0             0.0   
3             0.0      0.0                 0.0           0.0             0.0   
4             0.0      0.0                 0.0           0.0             0.0   

   Airport Service  Airport Terminal  American Restaurant  ...  \
0              0.0               0.0             0.000000  ...   
1              0.0               0.0             0.000000  ...   
2              0.0               0.0             0.000000  ...   
3              0.0               0.0             0.000000  ...   
4              0.0               0.0             0.043478  ...   

   Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                            0.0               0.0          0.0   
1                            0.0               0.0          0.0   
2                            0.0               0.0          0.0   
3                            0.0               0.0          0.0   
4                            0.0               0.0          0.0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wine Shop  Wings Joint  \
0                    0.0              0.0       0.0        0.0          0.0   
1                    0.0              0.0       0.0        0.0          0.0   
2                    0.0              0.0       0.0        0.0          0.0   
3                    0.0              0.0       0.0        0.0          0.0   
4                    0.0              0.0       0.0        0.0          0.0   

   Women's Store  Yoga Studio  
0            0.0          0.0  
1            0.0          0.0  
2            0.0          0.0  
3            0.0          0.0  
4            0.0          0.0  

[5 rows x 237 columns]

In [35]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [36]:
import numpy as np

In [37]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                                        Agincourt                Lounge   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North           Coffee Shop   
3                                  Bayview Village                  Café   
4                Bedford Park, Lawrence Manor East    Italian Restaurant   

       2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0  Latin American Restaurant        Breakfast Spot        Clothing Store   
1                        Gym          Skating Rink           Coffee Shop   
2                       Bank              Pharmacy         Deli / Bodega   
3                       Bank    Chinese Restaurant   Japanese Restaurant   
4                  Juice Bar           Coffee Shop        Sandwich Place   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0          Skating Rink           Yoga Studio    Dim Sum Restaurant   
1                  Pool                   Pub              Pharmacy   
2           Bridal Shop         Shopping Mall                 Diner   
3    Dim Sum Restaurant    Falafel Restaurant           Event Space   
4              Pharmacy      Sushi Restaurant     Indian Restaurant   

   8th Most Common Venue 9th Most Common Venue       10th Most Common Venue  
0   Ethiopian Restaurant     Electronics Store  Eastern European Restaurant  
1  General Entertainment               Gay Bar                      Dog Run  
2         Sandwich Place           Supermarket             Sushi Restaurant  
3   Ethiopian Restaurant     Electronics Store  Eastern European Restaurant  
4                   Café               Butcher              Thai Restaurant

In [38]:
# import k-means from clustering stage
from sklearn.cluster import KMeans

In [39]:
# set number of clusters
k_num_clusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=k_num_clusters, random_state=0).fit(toronto_grouped_clustering)
kmeans

KMeans(n_clusters=5, random_state=0)

In [40]:
kmeans.labels_[0:100]

array([3, 1, 3, 3, 3, 3, 3, 3, 3, 1, 4, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3,
       3, 1, 3, 3, 4, 3, 3, 3, 3, 3, 1, 3, 3, 3, 3, 3, 0, 2, 3, 3, 3, 3,
       1, 3, 3, 3, 1, 3, 0, 3, 0, 3, 3, 3, 1, 1, 2, 1, 3, 3, 3, 3, 3, 0,
       3, 1, 3, 1, 0, 3, 3, 3, 3, 3, 3, 3, 3, 1, 3, 4, 3, 3, 3, 3, 3, 3,
       4, 3, 3, 3, 3, 3, 4], dtype=int32)

In [41]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

In [42]:
toronto_merged = combined_data
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')
toronto_merged.head()

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.753259 -79.329656             3.0                  Park   
1  43.725882 -79.315572             3.0           Pizza Place   
2  43.654260 -79.360636             3.0           Coffee Shop   
3  43.718518 -79.464763             3.0        Clothing Store   
4  43.662301 -79.389494             3.0           Coffee Shop   

  2nd Most Common Venue  3rd Most Common Venue 4th Most Common Venue  \
0     Food & Drink Shop                  Hotel           Yoga Studio   
1           Coffee Shop  Portuguese Restaurant          Hockey Arena   
2                  Park                 Bakery               Theater   
3     Accessories Store  Vietnamese Restaurant              Boutique   
4      Sushi Restaurant            Yoga Studio   Fried Chicken Joint   

  5th Most Common Venue   6th Most Common Venue 7th Most Common Venue  \
0          Dessert Shop             Event Space  Ethiopian Restaurant   
1  Ethiopian Restaurant             Event Space     Electronics Store   
2        Breakfast Spot   Performing Arts Venue        Chocolate Shop   
3    Miscellaneous Shop  Furniture / Home Store           Event Space   
4                  Bank      Italian Restaurant                  Café   

         8th Most Common Venue        9th Most Common Venue  \
0            Electronics Store  Eastern European Restaurant   
1  Eastern European Restaurant             Department Store   
2                          Pub                   Restaurant   
3                  Coffee Shop               Cosmetics Shop   
4           Mexican Restaurant                        Diner   

  10th Most Common Venue  
0              Drugstore  
1              Drugstore  
2                   Café  
3   Ethiopian Restaurant  
4         Discount Store

In [43]:
toronto_merged_nonan = toronto_merged.dropna(subset=['Cluster Labels'])

In [44]:
import matplotlib.cm as cm
import matplotlib.colors as colors

In [45]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(k_num_clusters)
ys = [i + x + (i*x)**2 for i in range(k_num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged_nonan['Latitude'], toronto_merged_nonan['Longitude'], toronto_merged_nonan['Neighbourhood'], toronto_merged_nonan['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + '\n' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)]
        ).add_to(map_clusters)   
map_clusters

In [46]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 0, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
32       Scarborough             0.0            Playground   
50        North York             0.0          Intersection   
83   Central Toronto             0.0            Playground   
85       Scarborough             0.0            Playground   
91  Downtown Toronto             0.0                  Park   

     2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
32        Department Store           Event Space  Ethiopian Restaurant   
50  Furniture / Home Store                  Farm           Event Space   
83                   Trail      Department Store           Event Space   
85            Intersection                  Park           Event Space   
91              Playground                 Trail         Deli / Bodega   

   5th Most Common Venue        6th Most Common Venue  \
32     Electronics Store  Eastern European Restaurant   
50  Ethiopian Restaurant            Electronics Store   
83  Ethiopian Restaurant            Electronics Store   
85  Ethiopian Restaurant            Electronics Store   
91  Ethiopian Restaurant            Electronics Store   

          7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
32                    Drugstore            Donut Shop               Dog Run   
50  Eastern European Restaurant             Drugstore            Donut Shop   
83  Eastern European Restaurant             Drugstore            Donut Shop   
85  Eastern European Restaurant             Drugstore            Donut Shop   
91  Eastern European Restaurant             Drugstore            Donut Shop   

   10th Most Common Venue  
32    Distribution Center  
50                Dog Run  
83                Dog Run  
85                Dog Run  
91                Dog Run

In [47]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 1, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
10        North York             1.0           Pizza Place   
12       Scarborough             1.0                   Bar   
19      East Toronto             1.0      Asian Restaurant   
34        North York             1.0                   Bar   
37      West Toronto             1.0                   Bar   
63              York             1.0     Convenience Store   
65       Scarborough             1.0     Indian Restaurant   
75      West Toronto             1.0        Breakfast Spot   
77         Etobicoke             1.0        Sandwich Place   
87  Downtown Toronto             1.0       Airport Service   
93         Etobicoke             1.0           Pizza Place   
94         Etobicoke             1.0            Truck Stop   

    2nd Most Common Venue   3rd Most Common Venue 4th Most Common Venue  \
10       Sushi Restaurant                     Pub                Bakery   
12            Yoga Studio    Fast Food Restaurant                  Farm   
19      Health Food Store            Neighborhood                 Trail   
34         Massage Studio  Furniture / Home Store           Coffee Shop   
37  Vietnamese Restaurant        Asian Restaurant           Coffee Shop   
63         Breakfast Spot                 Brewery  Caribbean Restaurant   
65     Chinese Restaurant               Pet Store    Light Rail Station   
75              Gift Shop           Movie Theater      Cuban Restaurant   
77            Pizza Place       Mobile Phone Shop              Bus Line   
87       Airport Terminal                   Plane                   Bar   
93                    Gym            Skating Rink           Coffee Shop   
94          Garden Center               Drugstore                   Bar   

          5th Most Common Venue  6th Most Common Venue  \
10          Japanese Restaurant           Dessert Shop   
12           Falafel Restaurant            Event Space   
19                          Pub                Dog Run   
34        Vietnamese Restaurant          Metro Station   
37                  Yoga Studio                Theater   
63                  Pizza Place            Yoga Studio   
65       Furniture / Home Store  Vietnamese Restaurant   
75  Eastern European Restaurant                    Bar   
77         Ethiopian Restaurant      Electronics Store   
87             Sculpture Garden        Harbor / Marina   
93                         Pool                    Pub   
94          Rental Car Location            Yoga Studio   

          7th Most Common Venue  8th Most Common Venue  \
10         Ethiopian Restaurant      Electronics Store   
12         Ethiopian Restaurant      Electronics Store   
19           Dim Sum Restaurant                  Diner   
34         Caribbean Restaurant             Donut Shop   
37            Korean Restaurant             Beer Store   
63                        Diner            Event Space   
65           Dim Sum Restaurant                  Diner   
75                      Dog Run           Dessert Shop   
77  Eastern European Restaurant              Drugstore   
87                Boat or Ferry         Airport Lounge   
93                     Pharmacy  General Entertainment   
94                        Diner         Discount Store   

          9th Most Common Venue 10th Most Common Venue  
10  Eastern European Restaurant              Drugstore  
12  Eastern European Restaurant              Drugstore  
19               Discount Store    Distribution Center  
34               Discount Store    Distribution Center  
37                  Men's Store    Japanese Restaurant  
63         Ethiopian Restaurant      Electronics Store  
65               Discount Store    Distribution Center  
75           Italian Restaurant             Restaurant  
77                   Donut Shop                Dog Run  
87                 Airport Gate     Airport Food Court  
93                      Gay Bar                Dog Run  
94          Distribution Center     

In [48]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 2, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
57   North York             2.0        Baseball Field           Yoga Studio   
101   Etobicoke             2.0        Baseball Field           Yoga Studio   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
57   Fast Food Restaurant                  Farm    Falafel Restaurant   
101  Fast Food Restaurant                  Farm    Falafel Restaurant   

    6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
57            Event Space  Ethiopian Restaurant     Electronics Store   
101           Event Space  Ethiopian Restaurant     Electronics Store   

           9th Most Common Venue 10th Most Common Venue  
57   Eastern European Restaurant              Drugstore  
101  Eastern European Restaurant              Drugstore

In [49]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 3, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0          North York             3.0                  Park   
1          North York             3.0           Pizza Place   
2    Downtown Toronto             3.0           Coffee Shop   
3          North York             3.0        Clothing Store   
4    Downtown Toronto             3.0           Coffee Shop   
..                ...             ...                   ...   
96   Downtown Toronto             3.0            Restaurant   
97   Downtown Toronto             3.0                  Café   
99   Downtown Toronto             3.0           Coffee Shop   
100      East Toronto             3.0    Light Rail Station   
102         Etobicoke             3.0  Fast Food Restaurant   

    2nd Most Common Venue  3rd Most Common Venue 4th Most Common Venue  \
0       Food & Drink Shop                  Hotel           Yoga Studio   
1             Coffee Shop  Portuguese Restaurant          Hockey Arena   
2                    Park                 Bakery               Theater   
3       Accessories Store  Vietnamese Restaurant              Boutique   
4        Sushi Restaurant            Yoga Studio   Fried Chicken Joint   
..                    ...                    ...                   ...   
96                 Bakery                   Café    Italian Restaurant   
97             Restaurant            Coffee Shop    Seafood Restaurant   
99       Sushi Restaurant        Thai Restaurant      Ramen Restaurant   
100                  Park             Restaurant  Gym / Fitness Center   
102        Discount Store                 Bakery       Supplement Shop   

    5th Most Common Venue   6th Most Common Venue   7th Most Common Venue  \
0            Dessert Shop             Event Space    Ethiopian Restaurant   
1    Ethiopian Restaurant             Event Space       Electronics Store   
2          Breakfast Spot   Performing Arts Venue          Chocolate Shop   
3      Miscellaneous Shop  Furniture / Home Store             Event Space   
4                    Bank      Italian Restaurant                    Café   
..                    ...                     ...                     ...   
96            Coffee Shop     Japanese Restaurant    Caribbean Restaurant   
97         Sandwich Place                     Pub                  Bakery   
99         Ice Cream Shop       Indian Restaurant                   Diner   
100           Pizza Place              Comic Shop                 Brewery   
102         Burrito Place            Burger Joint  Thrift / Vintage Store   

           8th Most Common Venue        9th Most Common Venue  \
0              Electronics Store  Eastern European Restaurant   
1    Eastern European Restaurant             Department Store   
2                            Pub                   Restaurant   
3                    Coffee Shop               Cosmetics Shop   
4             Mexican Restaurant                        Diner   
..                           ...                          ...   
96          Taiwanese Restaurant                         Bank   
97                     Speakeasy                        Hotel   
99                    Restaurant          Japanese Restaurant   
100                Auto Workshop               Farmers Market   
102               Sandwich Place                   Kids Store   

    10th Most Common Venue  
0                Drugstore  
1                Drugstore  
2                     Café  
3     Ethiopian Restaurant  
4           Discount Store  
..                     ...  
96               Gift Shop  
97             Pizza Place  
99                    Café  
100   Fast Food Restaurant  
102          Tanning Salon  

[75 rows x 12 columns]

In [50]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 4, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
21        York             4.0                  Park         Women's Store   
35   East York             4.0                  Park     Convenience Store   
64        York             4.0                  Park     Convenience Store   
66  North York             4.0                  Park     Convenience Store   
98   Etobicoke             4.0                  Park                 River   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
21                  Pool      Department Store           Event Space   
35           Yoga Studio          Dessert Shop           Event Space   
64           Yoga Studio          Dessert Shop           Event Space   
66                   Bar           Yoga Studio    Dim Sum Restaurant   
98           Yoga Studio      Department Store           Event Space   

   6th Most Common Venue 7th Most Common Venue        8th Most Common Venue  \
21  Ethiopian Restaurant     Electronics Store  Eastern European Restaurant   
35  Ethiopian Restaurant     Electronics Store  Eastern European Restaurant   
64  Ethiopian Restaurant     Electronics Store  Eastern European Restaurant   
66    Falafel Restaurant           Event Space         Ethiopian Restaurant   
98  Ethiopian Restaurant     Electronics Store  Eastern European Restaurant   

   9th Most Common Venue       10th Most Common Venue  
21             Drugstore                   Donut Shop  
35             Drugstore                   Donut Shop  
64             Drugstore                   Donut Shop  
66     Electronics Store  Eastern European Restaurant  
98             Drugstore                   Donut Shop

We have successfully Segmented and Clustered Toronto Neighborhoods based on venue categories ...